New Term: Python `eval()` (#860)
Codecademy/docs: content/python/concepts/built-in-functions/terms/eval/eval.md

---

> Title: 'eval()'

Reviewer: Not sure if this standard has been changed but I believe we include the period: '.eval()'

Author: @caupolicandiaz: It's a function, not a method. No period.

Description: 'Returns the value of a Python expression passed as a string.'
Subjects:
  - 'Computer Science'
  - 'Data Science'
Tags:
  - 'Functions'
  - 'Methods'
  - 'Strings'
CatalogContent:
  - 'learn-python-3'
  - 'paths/data-science'
---

The **`eval()`** function returns the value of a python expression passed as a string.

## Syntax

```pseudo
eval(expression, globals, locals)
```

> The argument `expression` is the expression to evaluate. The optional argument `globals` is a dictionary defining what variables are in the `expression`'s global scope. If `globals` isn't specified, `eval()` uses the current global scope. The optional argument `locals` is a dictionary defining the variables in the `expression`'s local scope. If the `locals` argument is specified, the `globals` argument must be specified as well. It's not possible to use keyword arguments with `eval()`.

Reviewer 2: ```suggestion
The argument `expression` is the expression to evaluate. The optional argument `globals` is a [dictionary](https://www.codecademy.com/resources/docs/python/dictionaries) defining what variables are in the `expression`'s global [scope](https://www.codecademy.com/resources/docs/python/scope). If `globals` isn't specified, `eval()` uses the current global scope. The optional argument `locals` is a dictionary defining the variables in the `expression`'s local scope. If the `locals` argument is specified, the `globals` argument must be specified as well. The `eval()` function does not support keyword arguments.
```

> **Note:** While using the `globals` argument overrides the user defined variables avalible, if it doesn't specify a value for the key `__builtins__`, then a refrence for it is automatically added so that `eval()` will have access to all of Python's built-in names when evaluating `expression`.
> 
> **Note:** `eval()` doesn't work with compound statements or assignment operations. It only works with expressions that can be evaluated to be equal to some value.

Reviewer 2: ```suggestion
> **Note:** While using the `globals` argument overrides the user defined variables available, if it doesn't specify a value for the `__builtins__` key, then a reference for it is automatically added so that `eval()` will have access to all of Python's built-in names when evaluating `expression`.

> **Note:** `eval()` doesn't work with compound statements or assignment operations. It only works with expressions that can be evaluated to be equal to some value.
```

I'd previously spoken with @lucwhite about this, but I think we should try to embed any **Note:**s in a markdown quote, as suggested above along with other edits.

> **Note:** While using the `globals` argument overrides the user defined variables available, if it doesn't specify a value for the `__builtins__` key, then a reference for it is automatically added so that `eval()` will have access to all of Python's built-in names when evaluating `expression`.
>

> > **Note:** `eval()` doesn't work with compound statements or assignment operations. It only works with expressions that can be evaluated to be equal to some value.

Reviewer 3: ```suggestion
> **Note:** `eval()` does not support keyword arguments and it doesn't work on compound statements or assignment operations. It only works with expressions that can be evaluated to be equal to some value.
```

```py
x = 10
y = 5

print(eval("x + y")) # Output 15.
print(eval("x + y"), {"x":15, "y":y}) # Output 20.
```

## Security Concerns

> While `eval()` can be useful, care must be taken to understand the security implications of this function. If `eval()` is used with user generated strings, it's possible for a malicious user to execute arbitrary code through the function. Good programming practice generally advises against using `eval()`. If it is used, it should never be used with untrusted input.

Reviewer 2: I'm wondering if this entire section should be moved further up in the entry? Possibly somewhere in lines # 16-18?

> The following example uses `eval()` to evaluate an expression using variables in the current global scope, the evaluates the same expression with its own global scope:

Reviewer: ```suggestion
The following example uses `eval()` to evaluate an expression using variables in the current global scope, then evaluates the same expression with its own global scope:
```

> print(eval("x + y")) # Output 15.
> print(eval("x + y"), {"x":15, "y":y}) # Output 20.

Reviewer 2: ```suggestion
print(eval("x + y")) # Output: 15
print(eval("x + y"), {"x":15, "y":y}) # Output: 20
```

eval(expression, globals, locals)
```

- `expression`: The expression to evaluate.
- `globals` (optional): A [dictionary](https://www.codecademy.com/resources/docs/python/dictionaries) defining which [variables](https://www.codecademy.com/resources/docs/python/variables) are in the `expression`'s global [scope](https://www.codecademy.com/resources/docs/python/scope). If `globals` isn't specified, `eval()` uses the current global scope. 

> - `locals` (optional): A dictionary defining the variables in the `expression`'s local scope. If the `locals` argument is specified, the `globals` argument must be specified as well. 

Reviewer 2: ```suggestion
The `eval()` function uses the following parameters:

- The `expression` to evaluate.
- An optional `globals` [dictionary](https://www.codecademy.com/resources/docs/python/dictionaries) defining which [variables](https://www.codecademy.com/resources/docs/python/variables) are in the `expression`'s global [scope](https://www.codecademy.com/resources/docs/python/scope). If `globals` isn't specified, `eval()` uses the current global scope. 
- An optional `locals` dictionary defining the variables in the `expression`'s local scope. If the `locals` argument is specified, the `globals` argument must be specified as well. 
```
